### ISSUE TITLE
Theming

### ISSUE THREAD

**Issue by pngwn** [gradio-app/gradio#1488]
**Update**

I will use this as a tracking issue as we move forward with the theming implementation. This work needs to be split up. 

I think we have a good idea of the general usecases for theming from comments below, other community feedback, and internal discussion about how we want to empower the community to further customise (and share!) gradio apps. I will summarise as best I can in the dedicated issues. Broadly speaking these are: internal architecture, python API, creating themes, sharing themes.

- [x] #2353
- [x] #2354
- [x] #2355
- [x] #2356

**Context**

Generally speaking the addition of custom themes within gradio facilitates a few things which are increasingly important to us.

_Conceptually, this feature is deeply connected to 'custom components' which will be happening soon as well. Many of the benefits are the same, although they are technically separate features._

- **Expanding where Gradio is 'the right tool'**
  Popular gradio applications or power users of gradio hit a certain point where they want more control over the UI of gradio. Most of this control falls into the broad category of 'style', meaning that users want to customise the look and feel of their gradio apps either so they match branding in a broader sense (a gradio powered minisite needs to match the organisations branding) or within a more specific context (an embedded gradio app needs to match the style of the page it is contained within). Theming will allow gradio to 'continue to be the right tool' in these contexts which isn't really the case today, this extends the lifecycle of a gradio app as it grows in popularity or needs to be more tightly integrated into an existing context.
-  **Make it easier for us to ship new themes**
  We remove all of our other themes when we released 3.0 because they weren't up to the standard we wanted. A complete theming system it will make it easier for us to add new themes to gradio.  
- **Dark mode for every theme**
  We have never really had a good way to provide 'variants' of a theme, for example a dark and light mode for a single theme. A theming system will make this possible, allowing user to build production grade themes. User-created themes will be just as powerful as official ones (as they will be using the same underlying system). 
- **Empowering the community to self-serve**
  As gradio becomes more popular and the demands become more varied, gradio itself becomes less able to handle all of these usecases. Partly because of the gradio team's capacity to oversee the design and integration of new (and sometimes niche) features that further expand gradio's styling capabilities but also because the gradio team is way of adding too many niche features or APIs to gradio core. Theming will allow some of these niche theming requirements (which until now have often required additional API surface area and direct integration into gradio) to be handles in userland. This allows users to move faster (as gradio + the team stop being a bottleneck) but also facilitates changes that would not have made it into gradio core (due to being too niche)
- **Growing the community + ecosystem independently of gradio**
  The ability to build custom themes and share them with other users (who can also then use these user created themes in their apps) allows the community to make gradio more useful, interesting, and fun regardless of what gets added to gradio core. This has two additional benefits:
  - allowing the core team to work on expanding gradio's core functionality while minimising how much we get in the way of users building the applications they want to build.
  - allowing the community itself to grow and become more engaged with _one another_ by building and sharing new themes; and pushing the boundaries of what gradio is currently capable of, surfacing new and interesting usecases that could influence the direction of the library or future functionality.
- **Make working with custom components simpler**
  This obviously isn't a thing right now, but having a solid styling framework for gradio apps means that it will be much easier to create custom components that seamlessly (visually) integrate into the existing style of your gradio app, regardless of what theme you are using (official or community).

***

With 3.0 we removed the current themes + theming solution in favour of a more refined theming API in the future. It is still possible to achieve some level of customisation by use `Component.style()` kwargs or the global `css` override. But the global `css` over-ride is prone to breakage and the `style` method does not offer enough customisation (by design).

While there are a number of technical details that need to be ironed out (and i'll add them to this issue as I understand them more), this issue is primarily to gather thoughts, feedback, and use-cases for this theming API. 

Right now everything is on the table, but I'd like people to avoid suggesting technical solutions at this point. The two main questions i have are:

- What kind of use-cases people have that that theming would solve?
- What kind of API would make sense for this? What would 'feel good'?

Additionally, if anyone is aware of other python libraries that offer some level of theming, I'd love to hear about them and any experiences you have with them.

**Comment by hlydecker** [comment]
This isn't a python library, but an example of theming for a similar use case in the R world: https://rstudio.github.io/shinythemes/

The utility of the R shiny themes is primarily in quickly getting a decent looking application with minimal effort.

A use case for themes would be allowing user defined themes to be used, supporting things like corporate / university branding as well as just pure artistic expression.

**Comment by tylerlastovich** [comment]
I would like to use themes to tailor components so that they deeply integrate into pages/documents. This will enable the highly motivated to build production-quality apps. I perceive two distinct audiences of customization for Gradio; full control and templated theming. 

**Templated theming:** (most users)
- A lighter-weight style of customization with trapped variables (MD, CSS, JSON, YAML, etc)
- Font-familys, backgrounds, colors, syntax highlighting, sizing, borders, shadows
- Reusable/sharable templates for branding
- Perhaps a small WYSIWYG theme site that lets users select from a gallery, tune a few variables, then export a formed template?
- No UI experience needed

**Full customization:** (what I would like)
- The ability to style specific elements distinctly. (eliminates using :nth-child)
  - Ideally by assigning an element id to components
  - Global selectors for things like buttons, inputs
  - Could also apply to column, row, tab
- The ability to update the styles of components dynamically
- The ability to specify or handle breakpoints
- Requires knowledge of css

For durability, I think producing unique ids for components should be a high priority (even if not exposed to the user).

**Comment by pngwn** [comment]
> For durability, I think producing unique ids for components should be a high priority (even if not exposed to the user).

This is already supported. A number is used to create the unique ID but it isn't stable as you change the app. You can pass in a custom id which is stable. This is applied to the root element of a component.

**Comment by pngwn** [comment]
Thanks for all of the feedback, it is very useful!

I have updated the original issue now that this work is under way. It hopefully captures some of the benefits of theming and how we are thinking about theming internally. I've also made this a tracking issue, as we will be splitting this work out into parts (some of those parts have had very little thought thus far).

Feel free to continue to leave any feedback, in this issue. If your feedback relates to one specific piece of this work then feel free to comment in the dedicated issue too!
